crates/tinymist-world/src/lib.rs

```diff
@@ -200,3 +230,221 @@ pub fn parse_source_date_epoch(raw: &str) -> Result<DateTime<Utc>, String> {
+                Ok(response) => response,
+                Err(err) if matches!(err.status().map(|s| s.as_u16()), Some(404)) => {
+                    return Err(PackageError::NotFound(spec.clone()))
+                }
+                Err(err) => return Err(PackageError::NetworkFailed(Some(eco_format!("{err}")))),
+            };
+
+            let decompressed = flate2::read::GzDecoder::new(reader);
+            tar::Archive::new(decompressed)
+                .unpack(package_dir)
+                .map_err(|err| {
+                    std::fs::remove_dir_all(package_dir).ok();
+                    PackageError::MalformedArchive(Some(eco_format!("{err}")))
+                })
+        }, self.cert_path.as_deref())
+        .ok_or_else(|| PackageError::Other(Some(eco_format!("cannot spawn http thread"))))?
+    }
+}
+
+impl PackageRegistry for HttpsRegistry {
+    fn resolve(&self, spec: &PackageSpec) -> Result<std::sync::Arc<Path>, PackageError> {
+        self.prepare_package(spec)
+    }
+
+    fn packages(&self) -> &[(PackageSpec, Option<EcoString>)] {
+        self.packages.get_or_init(|| {
+            let url = "https://packages.typst.org/preview/index.json";
+
+            threaded_http(url, |resp| {
+                let reader = match resp.and_then(|r| r.error_for_status()) {
+                    Ok(response) => response,
+                    Err(err) => {
+                        // todo: silent error
+                        error!("Failed to fetch package index: {} from {}", err, url);
+                        return vec![];
+                    }
+                };
+
+                #[derive(serde::Deserialize)]
+                struct RemotePackageIndex {
+                    name: EcoString,
+                    version: PackageVersion,
+                    description: Option<EcoString>,
+                }
+
+                let index: Vec<RemotePackageIndex> = match serde_json::from_reader(reader) {
+                    Ok(index) => index,
+                    Err(err) => {
+                        error!("Failed to parse package index: {} from {}", err, url);
+                        return vec![];
+                    }
+                };
+
+                index
+                    .into_iter()
+                    .map(|e| {
+                        (
+                            PackageSpec {
+                                namespace: "preview".into(),
+                                name: e.name,
+                                version: e.version,
+                            },
+                            e.description,
+                        )
+                    })
+                    .collect::<Vec<_>>()
+            }, self.cert_path.as_deref())
+            .unwrap_or_default()
+        })
+    }
+}
+
+fn threaded_http<T: Send + Sync>(
+    url: &str,
+    f: impl FnOnce(Result<Response, reqwest::Error>) -> T + Send + Sync,
+    cert_path: Option<&Path>
```

Comment:
`f:` the callback argument should be the last argument, i.e. `cert_path:` should be placed before `f:`.

Reply:
I’ve adjusted the argument order so that `f` is now the last argument.